dns/stream_dialer.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+func resolveIP(ctx context.Context, resolver Resolver, rrType dnsmessage.Type, hostname string) ([]netip.Addr, error) {
+	ips := []netip.Addr{}
+	q, err := NewQuestion(hostname, rrType)
+	if err != nil {
+		return nil, err
+	}
+	response, err := resolver.Query(ctx, *q)
+	if err != nil {
+		return nil, err
+	}
+	if response.RCode != dnsmessage.RCodeSuccess {
+		return nil, fmt.Errorf("got %v (%d)", response.RCode.String(), response.RCode)
+	}
+	for _, answer := range response.Answers {
+		if answer.Header.Type != rrType {
+			continue
+		}
+		if rr, ok := answer.Body.(*dnsmessage.AResource); ok {
+			ips = append(ips, netip.AddrFrom4(rr.A))
+		}
+		if rr, ok := answer.Body.(*dnsmessage.AAAAResource); ok {
+			ips = append(ips, netip.AddrFrom16(rr.AAAA))
+		}
+	}
+	return ips, nil
+}
+
+// NewStreamDialer creates a [transport.StreamDialer] that uses Happy Eyeballs v2 to establish a connection.
+// It uses resolver to map host names to IP addresses, and the given dialer to attempt connections.
+func NewStreamDialer(resolver Resolver, dialer transport.StreamDialer) transport.StreamDialer {
```

Comment:
Shall we use `(transport.StreamDialer, error)` as the return type to align with other function (also in case in the future we might return errors, so the caller don't need to change the code). Even in this function, we can return errors if either `resolver` or `dialer` is `nil`.

Reply:
Good point. I've updated as per your suggestion